### ISSUE TITLE
Allowing scalar input (Jacobi theta)

### ISSUE THREAD

**Issue by miromarszal** [chualynn/Theta.jl#3]
Let me say first that your package saved my life :wink:  I might have a suggestion for an enhancement though.

I needed the Jacobi theta function, of which Riemann theta is a generalization. While wrapping variables into arrays works perfectly fine, it would be nice to have a dedicated method for scalar inputs in order to avoid allocating temporary 1-element arrays. 

I can imagine it could be too difficult due to how your implementation relies on inputs being arrays or it could bring negligible benefits in comparison with evaluation of the function itself. In such a case, a simple wrapper would be still very convenient.

**EDIT**

I was a bit fast with this one... I don't know what exactly happened, but a combination of kernel restarts and updates broke something and now I can't wrap scalars any more. Boiling your example down to single numbers results in:
```julia
M = hcat(0.794612+1.9986im)
R = RiemannMatrix(M)
z = [0.30657351+0.34017115im]
theta(z, RiemannMatrix(M))
```
```julia
MethodError: no method matching +(::Float64, ::Array{Float64,1})
For element-wise addition, use broadcasting with dot syntax: scalar .+ array
Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...) at operators.jl:538
  +(::Float64, ::Float64) at float.jl:401
  +(::AbstractFloat, ::Bool) at bool.jl:106
  ...
```
Here's my environment:
```julia
Pkg.status()
```
```
Status `~/.julia/environments/v1.5/Project.toml`
  [7073ff75] IJulia v1.23.2
  [76087f3c] NLopt v0.5.1
  [91a5bcdd] Plots v1.10.6
  [d330b81b] PyPlot v2.9.0
  [42bdb5c4] Theta v0.1.2
```
Any help appreciated!

**Comment by chualynn** [comment]
Thank you for the suggestion! I will look into writing a wrapper to allow scalar inputs.

As for the example that you gave, I think the problem is in the way that you define M. The following code works for me:
```julia
M = [0.794612+1.9986im]
R = RiemannMatrix(M)
z = [0.30657351+0.34017115im]
theta(z, RiemannMatrix(M))
```

**Comment by miromarszal** [comment]
Yes, that works for me too. I somehow thought I need to preserve the number of dimensions, hence the `hcat`. Thanks a lot!

**Comment by mytraya-gattu** [comment]
This, however, fails if I try to specify a characteristic.
`function θ(z, τ)
    M = [τ]
    R = RiemannMatrix(M)
    return theta([z], R, char=[1,1])
end
`
`
θ(randn(ComplexF64), 1.0im)
MethodError: no method matching +(::Vector{Float64}, ::Float64)
For element-wise addition, use broadcasting with dot syntax: array .+ scalar

Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...)
   @ Base operators.jl:578
  +(::T, ::T) where T<:Union{Float16, Float32, Float64}
   @ Base float.jl:408
  +(::Union{MathOptInterface.ScalarAffineFunction{T}, MathOptInterface.ScalarQuadraticFunction{T}}, ::T) where T
   @ MathOptInterface ~/.julia/packages/MathOptInterface/NCblk/src/Utilities/functions.jl:1783
  ...


Stacktrace:
 [1] theta(z::Vector{ComplexF64}, R::RiemannMatrix; char::Vector{Int64}, derivs::Vector{Any}, derivs_t::Vector{Any})
   @ Theta ~/.julia/packages/Theta/2Upgk/src/eval_theta.jl:42
 [2] θ(z::ComplexF64, τ::ComplexF64)
   @ Main ./In[51]:5
 [3] top-level scope
   @ In[55]:1

`

**Comment by mytraya-gattu** [comment]
A simple hack that I have found works is: char=[[1],[1]], and replacing additions in line  42 and line 61 (in eval_theta) with the vectorized version as the debugger suggests:
`
    x = real(z) + shift_x; # shift z by characteristic
`
to 
`
    x = real(z) .+ shift_x; # shift z by characteristic

`
And, 
`
    s = (2*π*im)^(nderivs) * sum([
(nderivs > 0 ? prod(transpose(d)*(p-w) for d in derivs) : 1) * exp(2*π*im*((0.5*(transpose(p-w)*R.X*(p-w))[1] + transpose(p-w)*x)[1]) - π*(transpose(p+w0)*R.Y*(p+w0))[1]) for p in E]); # compute sum over points in ellipsoid
`
to
`
    s = (2*π*im)^(nderivs) * sum([
(nderivs > 0 ? prod(transpose(d)*(p-w) for d in derivs) : 1) * exp(2*π*im*((0.5*(transpose(p-w)*R.X*(p-w))[1] .+ transpose(p-w)*x)[1]) - π*(transpose(p.+w0)*R.Y*(p.+w0))[1]) for p in E]); # compute sum over points in ellipsoid
`
However, I am unsure if this is the most efficient way to do this.
